research/object_detection/core/target_assigner.py

```diff
@@ -1897,3 +1899,170 @@ def assign_corner_offset_targets(
+             groundtruth_labels,
+             groundtruth_weights=None):
+    """Assign classification and regression targets to each box_pred.
+
+    For a given set of box_preds and groundtruth detections, match box_preds
+    to groundtruth_boxes and assign classification and regression targets to
+    each box_pred as well as weights based on the resulting match (specifying,
+    e.g., which box_preds should not contribute to training loss).
+
+    box_preds that are not matched to anything are given a classification target
+    of self._unmatched_cls_target which can be specified via the constructor.
+
+    Args:
+      box_preds: a BoxList representing N box_preds
+      groundtruth_boxes: a BoxList representing M groundtruth boxes
+      class_predictions: A tensor with shape [max_num_boxes, num_classes]
+        to be used by certain similarity calculators.
+      groundtruth_labels:  a tensor of shape [M, num_classes]
+        with labels for each of the ground_truth boxes. The subshape
+        [num_classes] can be empty (corresponding to scalar inputs).  When set
+        to None, groundtruth_labels assumes a binary problem where all
+        ground_truth boxes get a positive label (of 1).
+      groundtruth_weights: a float tensor of shape [M] indicating the weight to
+        assign to all box_preds match to a particular groundtruth box. The weights
+        must be in [0., 1.]. If None, all weights are set to 1. Generally no
+        groundtruth boxes with zero weight match to any box_preds as matchers are
+        aware of groundtruth weights. Additionally, `cls_weights` and
+        `reg_weights` are calculated using groundtruth weights as an added
+        safety.
+
+    Returns:
+      cls_targets: a float32 tensor with shape [num_box_preds, num_classes],
+        where the subshape [num_classes] is compatible with groundtruth_labels
+        which has shape [num_gt_boxes, num_classes].
+      cls_weights: a float32 tensor with shape [num_box_preds, num_classes],
+        representing weights for each element in cls_targets.
+      reg_targets: a float32 tensor with shape [num_box_preds, box_code_dimension]
+      reg_weights: a float32 tensor with shape [num_box_preds]
+
+    """
+    unmatched_class_label = tf.constant([1] + [0] * (groundtruth_labels.shape[1] - 1), tf.float32)
+
+    if groundtruth_weights is None:
+      num_gt_boxes = groundtruth_boxes.num_boxes_static()
+      if not num_gt_boxes:
+        num_gt_boxes = groundtruth_boxes.num_boxes()
+      groundtruth_weights = tf.ones([num_gt_boxes], dtype=tf.float32)
+
+    groundtruth_boxes.add_field(fields.BoxListFields.classes, groundtruth_labels)
+    box_preds.add_field(fields.BoxListFields.classes, class_predictions)
+
+    match_quality_matrix = self._similarity_calc.compare(
+        groundtruth_boxes,
+        box_preds)
+    match = self._matcher.match(match_quality_matrix,
+                                valid_rows=tf.greater(groundtruth_weights, 0))
```

Comment:
question - where do we ensure that the number of predicted boxes and number of padded groundtruth boxes is equal ?

Reply:
the default number of boxes to pad to in the config (max_number_of_boxes) is 100. I could explicitly set this in the config. I can also add an assert to confirm.